src/wamr/filesystem.cpp

```diff
@@ -1,20 +1,432 @@
+
+        ::iovec ioVecNative{
+            .iov_base = ADDR_APP_TO_NATIVE(ioVecBuffApp[i].buffOffset),
+            .iov_len = ioVecBuffApp[i].buffLen,
+        };
+
+        ioVecBuffNative[i] = ioVecNative;
+    }
+
+    // Write to fd
+    VALIDATE_NATIVE_ADDR(bytesWritten, sizeof(int32_t))
+    *bytesWritten =
+      ::writev(fileDesc.getLinuxFd(), ioVecBuffNative, ioVecCountApp);
+    if (*bytesWritten < 0) {
+        SPDLOG_ERROR(
+          "writev failed on fd {}: {}", fileDesc.getLinuxFd(), strerror(errno));
+    }
+
+    // Capture stdout if needed
+    conf::FaasmConfig& conf = conf::getFaasmConfig();
+    bool isStd = fd <= 2;
+    if (isStd && conf.captureStdout == "on") {
+        module->captureStdout(ioVecBuffNative, ioVecCountApp);
+    }
+
+    delete[] ioVecBuffNative;
+
+    return __WASI_ESUCCESS;
+}
+
+static int32_t wasi_path_create_directory(wasm_exec_env_t exec_env,
+                                          int32_t a,
+                                          int32_t* b,
+                                          char* c)
+{
+    SPDLOG_DEBUG("S - path_create_directory");
+    throw std::runtime_error("path_create_directory not implemented");
+    return 0;
+}
+
+static int32_t wasi_path_filestat_get(wasm_exec_env_t exec_env,
+                                      int32_t fd,
+                                      int32_t lookupFlags,
+                                      char* path,
+                                      int32_t pathLen,
+                                      __wasi_filestat_t* statApp)
+{
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    VALIDATE_NATIVE_ADDR(path, pathLen);
+    const std::string pathStr(path, pathLen);
+
+    SPDLOG_DEBUG("S - path_filestat_get {} ({})", fd, pathStr);
+
+    return doFileStat(fd, pathStr, statApp);
+}
+
+static int32_t wasi_path_link(wasm_exec_env_t exec_env,
+                              int32_t a,
+                              int32_t b,
+                              int32_t* c,
+                              char* d,
+                              int32_t e,
+                              int32_t* f,
+                              char* g)
+{
+    SPDLOG_DEBUG("S - path_link");
+    throw std::runtime_error("path_link not implemented");
+    return 0;
+}
+
+static int32_t wasi_path_open(wasm_exec_env_t exec_env,
+                              int32_t fdNative,
+                              int32_t lookupFlags,
+                              char* path,
+                              int32_t pathLen,
+                              int32_t openFlags,
+                              int64_t rightsBase,
+                              int64_t rightsInheriting,
+                              int32_t fdFlags,
+                              int32_t* fdApp)
+{
+    WAMRWasmModule* module = getExecutingWAMRModule();
+
+    VALIDATE_NATIVE_ADDR(path, pathLen);
+    const std::string pathStr(path, pathLen);
```

Comment:
This is an example of where I'm not sure about the use of `VALIDATE_NATIVE_ADDR`, first of all I don't think we do these sort of checks on strings passed to WAMR natives elsewhere (as I'm not sure they're necessary), but more importantly returning an `EBADF` code doesn't seem right when the path passed to `path_open` is invalid (as it's not related to a file descriptor), so could cause quite confusing issues.

Reply:
See the reply on the main `VALIDATE_NATIVE_ADDR` thread to see the points I make there.